src/lcdlib/lcdlib_common_oled.c

```diff
@@ -0,0 +1,531 @@
+    }
+
+    // Write the cleared buffer to the framebuffer device
+    lseek(fb_fd, 0, SEEK_SET);
+    ssize_t bytes_written = write(fb_fd, fb_buffer, OLED_WIDTH_BYTES * OLED_HEIGHT_BYTES);
+    if (bytes_written != OLED_WIDTH_BYTES * OLED_HEIGHT_BYTES)
+    {
+        print_err("Error: failed to write framebuffer content\n");
+        return LCD_ERR_WRITE;
+    }
+
+    return 0;
+}
+
+static int lcdlib_fb_write_string(int x, int y, const char *str, int str_len)
+{
+    if (fb_fd < 0 || !fb_buffer)
+    {
+        return -1;
+    }
+
+    print_err("2 Writing string to display at (%d, %d): %.*s\n", x, y, str_len, str);
+
+    // Read the current framebuffer content into the buffer
+    lseek(fb_fd, 0, SEEK_SET);
+    ssize_t bytes_read = read(fb_fd, fb_buffer, OLED_WIDTH_BYTES * OLED_HEIGHT_BYTES);
+    if (bytes_read != OLED_WIDTH_BYTES * OLED_HEIGHT_BYTES)
+    {
+        print_err("Error: failed to read framebuffer content\n");
+        free(fb_buffer);
+        fb_buffer = NULL;
+        close(fb_fd);
+        fb_fd = -1;
+        return LCD_ERR_READ;
+    }
+
+    int offset = y * OLED_WIDTH_PX / 2 + x / 2; // Calculate the position in the buffer
+
+    // Clear Line to be written:
+    lcdlib_clear_rows(y, LINE_HEIGHT_PX);
+
+    // 128 pixels in each line, each pixel is 4 bits (1 nibble)
+    for (int k = 0; k < str_len && *str; k++)
+    {
+        unsigned char glyph[CHARACTER_WIDTH_PX]; // 6x8 glyphs
+        load_glyph(*str, glyph);                 // Load glyph for character
+
+        for (int i = 0; i < CHARACTER_HEIGHT_PX; i++)
+        { // Iterate over rows of the glyph
+            for (int j = 0; j < CHARACTER_WIDTH_PX; j++)
+            {                                             // Iterate over columns of the glyph
+                int byte_index = offset + i * OLED_WIDTH_BYTES + j / 2; // Correct byte in the buffer
+                int pixel_value = (glyph[j] >> i) & 1 ? 0xF : 0x0;
+
+                if (j % 2 == 0)
+                {
+                    // Even column, modify the high nibble (4 bits)
+                    fb_buffer[byte_index] = (fb_buffer[byte_index] & 0x0F) | (pixel_value << 4);
+                }
+                else
+                {
+                    // Odd column, modify the low nibble (4 bits)
+                    fb_buffer[byte_index] = (fb_buffer[byte_index] & 0xF0) | pixel_value;
+                }
+            }
+        }
+
+        offset += CHARACTER_WIDTH_PX / 2; // Move to the next character position (6 pixels, 3 bytes)
+        str++;
+    }
+
+    // Write the buffer to the framebuffer device
+    lseek(fb_fd, 0, SEEK_SET);
+    ssize_t bytes_written = write(fb_fd, fb_buffer, OLED_WIDTH_BYTES * OLED_HEIGHT_BYTES);
+    if (bytes_written != OLED_WIDTH_BYTES * OLED_HEIGHT_BYTES)
+    {
+        print_err("Error: failed to write framebuffer content\n");
+        return LCD_ERR_WRITE;
+    }
+
+    return 0;
+}
+
+int lcdlib_open_dev(void)
+{
+    fb_fd = open("/dev/fb0", O_RDWR);
```

Comment:
pls can we make this an argument?
if we have two displays, would it be fb0 and fb1?

Reply:
Not sure how to go about the architecture; since both screens share the same spi lines and chip select and the path is simply directed by a mux. Because of this, they will be one device from the linux point of view. For now, I have focused on just 1 screen enablement and I will do a write up with ideas on how to enhance in the future for 2 screens. So is it ok to keep as fb0 for now?

Reply:
Understood. Thanks.